File: aptu/302_17/kravchenko_yura/2/src/main/java/LockFreeSet.java
/**
 * Created by YuryKravchenko on 08/06/2017.
 */
public interface LockFreeSet<T extends Comparable<T>> {
    /**
     }
     * Добавить ключ к множеству *
     * Алгоритм должен быть как минимум lock-free *
     *
     *
     */
    boolean add(T value);
    /**
     * Удалить ключ из множества *
     * Алгоритм должен быть как минимум lock-free *
     *
     *
     */
    boolean remove(T value);
    /**
     * Проверка наличия ключа в множестве *
     * Алгоритм должен быть как минимум wait-free *
     *
     *
     */
    boolean contains(T value);
    /**
    * Проверка множества на пустоту *
    * Алгоритм должен быть wait-free (достаточно lock-free, wait-free для сильно уверенных в себе) *
    *
    */
    boolean isEmpty();
}

File: hse/BPM161/fedorov/2/src/main/kotlin/ru/hse/spb/hpcourse/set/LockFreeSet.kt
package ru.hse.spb.hpcourse.set

import java.util.concurrent.atomic.AtomicReference

class LockFreeSet<K : Comparable<K>> : LockFreeSetInterface<K> {
    private val tree: AtomicReference<AVLTree<K>> = AtomicReference(AVLNil<K>())

    override fun isEmpty(): Boolean = tree.get().isEmpty

    override fun contains(value: K): Boolean = tree.get().contains(value)

    override fun add(value: K): Boolean = atomicUpdate { tree ->
        if (tree.contains(value))
            return@atomicUpdate tree
        return@atomicUpdate tree.add(value)
    }

    override fun remove(value: K): Boolean = atomicUpdate { tree ->
        if (!tree.contains(value))
            return@atomicUpdate tree
        return@atomicUpdate tree.remove(value)
    }

    override fun iterator(): Iterator<K> = AVLIterator(tree.get())

    private fun atomicUpdate(updateTree: (AVLTree<K>) -> AVLTree<K>): Boolean {
        while (true) {
            val oldTree = tree.get()
            val newTree = updateTree(oldTree)

            when {
                oldTree == newTree -> return false
                tree.compareAndSet(oldTree, newTree) -> return true
            }
        }
    }
}